services/heartbeat/etcd/store.go

```diff
@@ -142,109 +168,93 @@ func (c *client) Heartbeat(service, instance string, ttl time.Duration) error {
+
+func (c *client) get(key string) ([]string, error) {
 	ctx, cancel := c.context()
 	defer cancel()
 
-	gr, err := c.kv.Get(ctx, servicePrefix(service), clientv3.WithPrefix())
+	gr, err := c.kv.Get(ctx, key, clientv3.WithPrefix())
 	if err != nil {
 		c.m.etcdGetError.Inc(1)
 		return nil, err
 	}
 
 	r := make([]string, len(gr.Kvs))
 	for i, kv := range gr.Kvs {
-		r[i] = instanceFromKey(string(kv.Key), service)
+		r[i] = instanceFromKey(string(kv.Key), key)
 	}
 	return r, nil
 }
 
 func (c *client) Watch(service string) (xwatch.Watch, error) {
-	c.Lock()
-	watchable, ok := c.watchables[service]
+	serviceKey := servicePrefix(service)
 
+	c.Lock()
+	watchable, ok := c.watchables[serviceKey]
 	if !ok {
-		watchChan := c.watchChan(service)
-		c.m.etcdWatchCreate.Inc(1)
-
 		watchable = xwatch.NewWatchable()
-		c.watchables[service] = watchable
-
-		go func() {
-			ticker := time.Tick(c.opts.WatchChanCheckInterval())
-
-			for {
-				select {
-				case r, ok := <-watchChan:
-					if !ok {
-						c.logger.Warnf("etcd watch channel closed on service %s, recreating a watch channel", service)
-
-						// avoid recreating watch channel too frequently
-						time.Sleep(c.opts.WatchChanResetInterval())
-
-						watchChan = c.watchChan(service)
-						c.m.etcdWatchReset.Inc(1)
-
-						continue
-					}
-					// handle the update
-					if r.Err() != nil {
-						c.logger.Errorf("received error on watch channel: %v", r.Err())
-						c.m.etcdWatchError.Inc(1)
-						// do not continue here, even though the update contains an error
-						// we still take this chance to attemp a Get() for the latest value
-					}
-
-					// we retry because if Get() failed on an watch update,
-					// it has to wait 10 mins to be notified to try again
-					if err := c.retrier.Attempt(func() error {
-						return c.update(watchable, service)
-					}); err != nil {
-						c.logger.Errorf("received updates for service %s, but failed to get value: %v", service, err)
-					}
-				case <-ticker:
-					c.RLock()
-					numWatches := watchable.NumWatches()
-					c.RUnlock()
-
-					if numWatches != 0 {
-						// there are still watches on this watchable, do nothing
-						continue
-					}
-
-					if cleanedUp := c.tryCleanUp(service); cleanedUp {
-						return
-					}
-				}
-			}
-		}()
+		c.watchables[serviceKey] = watchable
+
+		go c.wh.Run(serviceKey)
 	}
 	c.Unlock()
+
 	_, w, err := watchable.Watch()
 	return w, err
 }
 
-func (c *client) watchChan(service string) clientv3.WatchChan {
-	return c.watcher.Watch(
-		context.Background(),
-		servicePrefix(service),
-		// WithPrefix so that the watch will receive any changes
-		// from the instances under the service
-		clientv3.WithPrefix(),
-		// periodically (appx every 10 mins) checks for the latest data
-		// with or without any update notification
-		clientv3.WithProgressNotify(),
-		// receive initial notification once the watch channel is created
-		clientv3.WithCreatedNotify(),
+func (c *client) update(key string) error {
```

Comment:
yeah... since Golang does not have generic, this very similar code has to be duplicated, because one store is managing kv.ValueWatchable and the other one is managing watch.Watchable. Also the logic is slightly different. the kv store cares about versioning and the other one don't, thus the separation